deploy/containers/credhub_setup/auth.go

```diff
@@ -0,0 +1,120 @@
+package main
+
+// This file contains code to deal with UAA / authentication.
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"time"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+)
+
+func makeHTTPClientWithCA(serverName string, caCert []byte) (*http.Client, error) {
+	certPool, err := x509.SystemCertPool()
+	if err != nil {
+		return nil, fmt.Errorf("could not get system cert pool: %w", err)
+	}
+	ok := certPool.AppendCertsFromPEM(caCert)
+	if !ok {
+		fmt.Printf("Warning: failed to add CA certs for %s\n", serverName)
+	}
+	return &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs:    certPool,
+				ServerName: serverName,
+			},
+		},
+		Timeout: 60 * time.Second,
+	}, nil
+}
+
+// authenticate with UAA, returning the access token and refresh token.
+func authenticate(ctx context.Context, clientID, clientSecret string) (*http.Client, error) {
+	link, err := getCCLinkData(ctx)
```

Comment:
`getCCLinkData` is a hidden dependency.

Reply:
I don't think the resolution of the link should happen magically inside of this function. The link information (as separate InternalServiceHostname and CACert properties) should be passed to either the constructor of the struct or via the parameter list of the method.

Reply:
Moved things around so that the token URL, and other things are all passed in.  Re-reading the comments, though, I suspect I should be passing in the UAA CA cert as well; I should do that…